{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d9a8f8c7-adb0-4164-9476-9e8f748e5eb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# regenerate the keys\n",
    "import os, sys\n",
    "os.system('python3 gen.py')\n",
    "sys.argv[1] = 'fam_unknown.fasta'\n",
    "sys.argv[2] = 'regen.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cab3dbcf-256e-4fba-9e4c-accbc97a7e43",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib64/python3.11/site-packages/skbio/alignment/_pairwise.py:595: EfficiencyWarning: You're using skbio's python implementation of Needleman-Wunsch alignment. This is known to be very slow (e.g., thousands of times slower than a native C implementation). We'll be adding a faster version soon (see https://github.com/biocore/scikit-bio/issues/254 to track progress on this).\n",
      "  warn(\"You're using skbio's python implementation of Needleman-Wunsch \"\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "import matplotlib.pyplot as plt  # type: ignore\n",
    "import functools\n",
    "import skbio  # type: ignore\n",
    "import scipy as sp  # type: ignore\n",
    "from scipy import cluster\n",
    "import collections\n",
    "from typing import Callable, Any, Optional, Tuple, TYPE_CHECKING\n",
    "if TYPE_CHECKING:\n",
    "    import skbio.SequenceCollection  # type: ignore\n",
    "\n",
    "import sys\n",
    "from skbio.io import read  # type: ignore\n",
    "from skbio import DNA, TabularMSA\n",
    "\n",
    "# print('Take in the fasta file via a filename argument 1, and output the answers the filename provided in argument 2.')\n",
    "input_file_path = sys.argv[1]\n",
    "output_file_path = sys.argv[2]\n",
    "\n",
    "# copied from skbio cookbook\n",
    "query_sequences = []\n",
    "sequences = collections.OrderedDict()\n",
    "for seq in skbio.io.read(input_file_path, format=\"fasta\"):\n",
    "    seq_id = seq.metadata[\"description\"]\n",
    "    # seq_tax = ' '.join(taxonomies[seq.metadata['id']].split('; ')[:display_tax_levels])\n",
    "    # seq_id = '%s (%s)' % (seq.metadata['id'], seq_tax)\n",
    "    sequences[seq_id] = DNA(sequence=seq, metadata={\"id\": seq_id})\n",
    "    query_sequences.append(DNA(sequence=seq, metadata={\"id\": seq_id}))\n",
    "\n",
    "\n",
    "def kmer_distance(sequence1: skbio.Sequence, sequence2: skbio.Sequence, k: int =3, overlap: bool=True) -> float:\n",
    "    \"\"\"Compute the kmer distance between a pair of sequences\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    sequence1 : skbio.Sequence\n",
    "    sequence2 : skbio.Sequence\n",
    "    k : int, optional\n",
    "        The word length.\n",
    "    overlapping : bool, optional\n",
    "        Defines whether the k-words should be overlapping or not\n",
    "        overlapping.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    float\n",
    "        Fraction of the set of k-mers from both sequence1 and\n",
    "        sequence2 that are unique to either sequence1 or\n",
    "        sequence2.\n",
    "\n",
    "    Raises\n",
    "    ------\n",
    "    ValueError\n",
    "        If k < 1.\n",
    "\n",
    "    Notes\n",
    "    -----\n",
    "    k-mer counts are not incorporated in this distance metric.\n",
    "\n",
    "    \"\"\"\n",
    "    sequence1_kmers = set(map(str, sequence1.iter_kmers(k=k, overlap=overlap)))\n",
    "    sequence2_kmers = set(map(str, sequence2.iter_kmers(k=k, overlap=overlap)))\n",
    "    all_kmers = sequence1_kmers | sequence2_kmers\n",
    "    shared_kmers = sequence1_kmers & sequence2_kmers\n",
    "    number_unique = len(all_kmers) - len(shared_kmers)\n",
    "    fraction_unique = number_unique / len(all_kmers)\n",
    "    return fraction_unique\n",
    "\n",
    "# The recursive function:\n",
    "def progressive_msa(sequences: \"skbio.SequenceCollection\", pairwise_aligner: skbio.alignment.global_pairwise_align_nucleotide, guide_tree: skbio.TreeNode=None) -> skbio.TabularMSA:\n",
    "    \"\"\"Perform progressive msa of sequences\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    sequences : skbio.SequenceCollection\n",
    "        The sequences to be aligned.\n",
    "    metric : function, optional\n",
    "      Function that returns a single distance value when given a pair of\n",
    "      skbio.Sequence objects. This will be used to build a guide tree if one\n",
    "      is not provided.\n",
    "    guide_tree : skbio.TreeNode, optional\n",
    "        The tree that should be used to guide the alignment process.\n",
    "    pairwise_aligner : function\n",
    "        Function that should be used to perform the pairwise alignments,\n",
    "        for example skbio.alignment.global_pairwise_align_nucleotide. Must\n",
    "        support skbio.Sequence objects or skbio.TabularMSA objects\n",
    "        as input.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    skbio.TabularMSA\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    if guide_tree is None:\n",
    "        guide_dm = skbio.DistanceMatrix.from_iterable(\n",
    "            iterable=sequences, metric=kmer_distance, key=\"id\"\n",
    "        )\n",
    "        guide_lm = sp.cluster.hierarchy.average(y=guide_dm.condensed_form())\n",
    "        guide_tree = skbio.TreeNode.from_linkage_matrix(\n",
    "            linkage_matrix=guide_lm, id_list=guide_dm.ids\n",
    "        )\n",
    "\n",
    "    seq_lookup = {s.metadata[\"id\"]: s for i, s in enumerate(sequences)}\n",
    "\n",
    "    # working our way down, first children may be super-nodes,\n",
    "    # then eventually, they'll be leaves\n",
    "    c1, c2 = guide_tree.children\n",
    "\n",
    "    # Recursive base case\n",
    "    if c1.is_tip():\n",
    "        c1_aln = seq_lookup[c1.name]\n",
    "    else:\n",
    "        c1_aln = progressive_msa(\n",
    "            sequences=sequences, pairwise_aligner=pairwise_aligner, guide_tree=c1\n",
    "        )\n",
    "\n",
    "    if c2.is_tip():\n",
    "        c2_aln = seq_lookup[c2.name]\n",
    "    else:\n",
    "        c2_aln = progressive_msa(\n",
    "            sequences=sequences, pairwise_aligner=pairwise_aligner, guide_tree=c2\n",
    "        )\n",
    "\n",
    "    # working our way up, doing alignments, from the bottom up\n",
    "    alignment, _, _ = pairwise_aligner(seq1=c1_aln, seq2=c2_aln)\n",
    "\n",
    "    # this is a temporary hack as the aligners in skbio 0.4.1 are dropping\n",
    "    # metadata - this makes sure that the right metadata is associated with\n",
    "    # the sequence after alignment\n",
    "    if isinstance(c1_aln, skbio.Sequence):\n",
    "        alignment[0].metadata = c1_aln.metadata\n",
    "        len_c1_aln = 1\n",
    "    else:\n",
    "        for i in range(len(c1_aln)):\n",
    "            alignment[i].metadata = c1_aln[i].metadata\n",
    "        len_c1_aln = len(c1_aln)\n",
    "    if isinstance(c2_aln, skbio.Sequence):\n",
    "        alignment[1].metadata = c2_aln.metadata\n",
    "    else:\n",
    "        for i in range(len(c2_aln)):\n",
    "            alignment[len_c1_aln + i].metadata = c2_aln[i].metadata\n",
    "\n",
    "    # feed alignment back up, for further aligment, or eventually final return\n",
    "    return alignment\n",
    "\n",
    "\n",
    "def progressive_msa_and_tree(\n",
    "    sequences: \"skbio.SequenceCollection\",\n",
    "    pairwise_aligner: skbio.alignment.global_pairwise_align_nucleotide,\n",
    "    guide_tree: skbio.TreeNode=None,\n",
    "    display_aln: bool=False,\n",
    "    display_tree: bool=False,\n",
    ") -> Tuple[skbio.alignment, skbio.TabularMSA]:\n",
    "    \"\"\"Perform progressive msa of sequences and build a UPGMA tree\n",
    "    Parameters\n",
    "    ----------\n",
    "    sequences : skbio.SequenceCollection\n",
    "        The sequences to be aligned.\n",
    "    pairwise_aligner : function\n",
    "        Function that should be used to perform the pairwise alignments,\n",
    "        for example skbio.alignment.global_pairwise_align_nucleotide. Must\n",
    "        support skbio.Sequence objects or skbio.TabularMSA objects\n",
    "        as input.\n",
    "    metric : function, optional\n",
    "      Function that returns a single distance value when given a pair of\n",
    "      skbio.Sequence objects. This will be used to build a guide tree if one\n",
    "      is not provided.\n",
    "    guide_tree : skbio.TreeNode, optional\n",
    "        The tree that should be used to guide the alignment process.\n",
    "    display_aln : bool, optional\n",
    "        Print the alignment before returning.\n",
    "    display_tree : bool, optional\n",
    "        Print the tree before returning.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    skbio.alignment\n",
    "    skbio.TreeNode\n",
    "\n",
    "    \"\"\"\n",
    "    msa = progressive_msa(\n",
    "        sequences=sequences, pairwise_aligner=pairwise_aligner, guide_tree=guide_tree\n",
    "    )\n",
    "\n",
    "    if display_aln:\n",
    "        print(msa)\n",
    "\n",
    "    msa_dm = skbio.DistanceMatrix.from_iterable(iterable=msa, metric=kmer_distance, key=\"id\")\n",
    "    msa_lm = sp.cluster.hierarchy.average(y=msa_dm.condensed_form())\n",
    "    msa_tree = skbio.TreeNode.from_linkage_matrix(\n",
    "        linkage_matrix=msa_lm, id_list=msa_dm.ids\n",
    "    )\n",
    "    if display_tree:\n",
    "        print(\"\\nOutput tree:\")\n",
    "        d = sp.cluster.hierarchy.dendrogram(\n",
    "            msa_lm,\n",
    "            labels=msa_dm.ids,\n",
    "            orientation=\"right\",\n",
    "            link_color_func=lambda x: \"black\",\n",
    "        )\n",
    "    return msa, msa_tree\n",
    "\n",
    "\n",
    "# taken from lecture 4\n",
    "global_pairwise_align_nucleotide = functools.partial(\n",
    "    skbio.alignment.global_pairwise_align_nucleotide, penalize_terminal_gaps=True\n",
    ")\n",
    "\n",
    "msa, tree = progressive_msa_and_tree(\n",
    "    sequences=query_sequences,\n",
    "    pairwise_aligner=global_pairwise_align_nucleotide,\n",
    "    display_tree=False,\n",
    "    display_aln=False,\n",
    ")\n",
    "\n",
    "msa_dm = skbio.DistanceMatrix.from_iterable(\n",
    "    iterable=msa, metric=skbio.sequence.distance.hamming, key=\"id\"\n",
    ")\n",
    "\n",
    "\n",
    "# get how far away each of them is from 'Sequence_You', order them\n",
    "ordered = sorted(list(zip(msa_dm[msa_dm.ids.index(\"Sequence_You\")], msa_dm.ids)))\n",
    "\n",
    "# convert them to a dictionary converting their sequence number to whether they are a grandparent or not\n",
    "def convert_index_to_lineage(index: int) -> str:\n",
    "    if index == 0:\n",
    "        return \"Sequence_You\"\n",
    "    elif index <= 2:\n",
    "        return \"Sequence_Parent\"\n",
    "    elif index > 2 and index <= 6:\n",
    "        return \"Sequence_Grandparent\"\n",
    "    else:\n",
    "        return \"Index_Too_Big\"\n",
    "\n",
    "\n",
    "ordered_lineage = {\n",
    "    t[1]: convert_index_to_lineage(index) for index, t in enumerate(ordered)\n",
    "}\n",
    "\n",
    "# map the new names onto the file, export to output file name\n",
    "# print(ordered_lineage[seq.metadata['description']])\n",
    "with open(output_file_path, \"w\") as output_file:\n",
    "    for seq in skbio.io.read(input_file_path, format=\"fasta\"):\n",
    "        seq.metadata[\"description\"] = ordered_lineage[seq.metadata[\"description\"]]\n",
    "        seq.write(output_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3a06544-5354-4f8a-b827-94f2c293ef84",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('diff fam_key.fasta regen.fasta')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
